File: backend/fund-service/source/core/fund_manager.py
# backend/fund-service/source/core/fund_manager.py
import logging
import uuid
from typing import Dict, Any, List, Optional

from source.models.fund import Fund

from source.db.fund_repository import FundRepository
from source.core.crypto_manager import CryptoManager

from source.utils.metrics import track_fund_created

logger = logging.getLogger('fund_manager')

class FundManager:
    """Manager for fund operations"""

    def __init__(
            self,
            fund_repository: FundRepository,
            crypto_manager: CryptoManager,
    ):
        """Initialize the fund manager with dependencies"""
        self.fund_repository = fund_repository
        self.crypto_manager = crypto_manager

    async def create_fund(self, fund_data: Dict[str, Any], user_id: str) -> Dict[str, Any]:
        """
        Create a new fund for a user with blockchain wallet integration
        
        Args:
            fund_data: Fund data dictionary
            user_id: User ID
            
        Returns:
            Result dictionary with success flag and fund_id
        """
        logger.info(f"Creating fund for user {user_id}")
        
        # Check if the user already has a fund (one-to-one relationship)
        existing_fund = await self.fund_repository.check_fund_exists(user_id)
        if existing_fund:
            return {
                "success": False,
                "error": "User already has a fund profile"
            }
        
        # Create fund model
        try:
            fund = Fund(
                user_id=user_id,
                fund_id=str(uuid.uuid4())
            )
            
            # Extract properties from fund data
            properties = {}
            
            # Map general properties to the properties object
            property_fields = [
                'fundName', 'legalStructure', 'location', 'yearEstablished', 'aumRange',
                'profilePurpose', 'otherPurposeDetails', 'investmentStrategy'
            ]
            
            for field in property_fields:
                if field in fund_data:
                    properties[field] = fund_data[field]

            # Process team members
            team_members = []
            if 'team_members' in fund_data and isinstance(fund_data['team_members'], list):
                for idx, member in enumerate(fund_data['team_members']):
                    team_member = {
                        'order': idx,
                        'firstName': member.get('firstName', ''),
                        'lastName': member.get('lastName', ''),
                        'birthDate': member.get('birthDate', ''),
                        'role': member.get('role', ''),
                        'yearsExperience': member.get('yearsExperience', ''),
                        'currentEmployment': member.get('currentEmployment', ''),
                        'investmentExpertise': member.get('investmentExpertise', ''),
                        'linkedin': member.get('linkedin', ''),
                        'education': member.get('education', '')
                    }
                    team_members.append(team_member)
            
            # Save fund to database with properties and team members
            result = await self.fund_repository.create_fund_with_details(
                fund.to_dict(),
                properties,
                team_members
            )

            if result and result.get("success"):
                fund_id = result['fund_id']
                
                # STEP 1: Create blockchain wallet for the fund
                logger.info(f"Creating blockchain wallet for fund {fund_id}")
                wallet_result = await self.crypto_manager.create_wallet(user_id, fund_id)
                
                if not wallet_result.get("success"):
                    logger.error(f"Failed to create wallet for fund {fund_id}: {wallet_result.get('error')}")
                    # Note: We could rollback the fund creation here if desired
                    return {
                        "success": False,
                        "error": f"Fund created but wallet creation failed: {wallet_result.get('error')}"
                    }
                
                logger.info(f"Blockchain wallet created successfully for fund {fund_id}")
                
                # Track metrics
                track_fund_created(user_id)
                
                return {
                    "success": True,
                    "fund_id": fund_id,
                    "wallet_address": wallet_result.get("address")
                }
            else:
                return {
                    "success": False,
                    "error": result.get("error", "Failed to save fund")
                }
            
        except Exception as e:
            logger.error(f"Error creating fund: {e}")
            return {
                "success": False,
                "error": f"Error creating fund: {str(e)}"
            }
    
    async def get_fund(self, user_id: str) -> Dict[str, Any]:
        """
        Get a fund for a user with blockchain wallet information
        
        Args:
            user_id: User ID
            
        Returns:
            Result dictionary with success flag and fund data
        """
        logger.info(f"Getting fund for user {user_id}")
        
        try:
            fund = await self.fund_repository.get_fund_by_user(user_id)

            if not fund:
                return {
                    "success": False,
                    "error": "Fund not found",
                    "fund": None
                }
            
            fund_id = fund['fund_id']
            
            # Get blockchain wallet information
            wallet_data = await self.crypto_manager.get_wallet(user_id, fund_id)
            
            if wallet_data:
                # Add wallet information to fund data
                fund['wallet'] = {
                    'address': wallet_data.get('address'),
                    'active_at': wallet_data.get('active_at')
                }
                logger.info(f"Retrieved fund with wallet for user {user_id}")
            else:
                logger.warning(f"Fund found but no wallet data for user {user_id}")
                fund['wallet'] = None
            
            return {
                "success": True,
                "fund": fund
            }
            
        except Exception as e:
            logger.error(f"Error getting fund for user {user_id}: {e}")
            return {
                "success": False,
                "error": f"Error getting fund: {str(e)}",
                "fund": None
            }
    
    async def update_fund(self, fund_data: Dict[str, Any], user_id: str) -> Dict[str, Any]:
        """
        Update a fund for a user (blockchain wallet remains unchanged)
        
        Args:
            fund_data: Fund data dictionary with updates
            user_id: User ID
            
        Returns:
            Result dictionary with success flag
        """
        logger.info(f"Updating fund for user {user_id}")
        
        try:
            # First, get the fund to verify existence and get fund_id
            current_fund = await self.fund_repository.get_fund_by_user(user_id)
            
            if not current_fund:
                return {
                    "success": False,
                    "error": "Fund not found"
                }
            
            fund_id = current_fund['fund_id']
            logger.info(f"Found fund with ID: {fund_id}")
            
            # Prepare update data
            update_object = {}
                        
            # Handle properties
            properties = {}
            
            # Check if any property fields are being updated
            property_fields = [
                'fundName', 'legalStructure', 'location', 'yearEstablished', 'aumRange',
                'profilePurpose', 'otherPurposeDetails', 'investmentStrategy'
            ]
            
            for field in property_fields:
                if field in fund_data:
                    properties[field] = fund_data[field]
            
            if properties:
                update_object['properties'] = properties
            
            # Handle team members
            if 'team_members' in fund_data and isinstance(fund_data['team_members'], list):
                team_members = []
                
                for idx, member in enumerate(fund_data['team_members']):
                    team_member = {
                        'order': idx,
                        'firstName': member.get('firstName', ''),
                        'lastName': member.get('lastName', ''),
                        'birthDate': member.get('birthDate', ''),
                        'role': member.get('role', ''),
                        'yearsExperience': member.get('yearsExperience', ''),
                        'currentEmployment': member.get('currentEmployment', ''),
                        'investmentExpertise': member.get('investmentExpertise', ''),
                        'linkedin': member.get('linkedin', ''),
                        'education': member.get('education', '')
                    }
                    team_members.append(team_member)
                
                if team_members:
                    update_object['team_members'] = team_members
            
            # Apply the updates
            if update_object:
                success = await self.fund_repository.update_fund(fund_id, user_id, update_object)
                
                # Note: Wallet operations are not needed for fund updates
                # The blockchain wallet remains the same throughout the fund's lifecycle
                
                if success:
                    logger.info(f"Successfully updated fund {fund_id}")
                    return {
                        "success": True
                    }
                else:
                    logger.error(f"Failed to update fund {fund_id}")
                    return {
                        "success": False,
                        "error": "Failed to update fund"
                    }
            else:
                logger.info("No valid updates provided")
                return {
                    "success": True,
                    "message": "No valid updates provided"
                }
                
        except Exception as e:
            logger.error(f"Error updating fund for user {user_id}: {e}")
            return {
                "success": False,
                "error": f"Error updating fund: {str(e)}"
            }

    async def get_fund_wallet_info(self, user_id: str) -> Dict[str, Any]:
        """
        Get wallet information for a user's fund
        
        Args:
            user_id: User ID
            
        Returns:
            Result dictionary with wallet information
        """
        try:
            # First get the fund to get the fund_id
            fund = await self.fund_repository.get_fund_by_user(user_id)
            
            if not fund:
                return {
                    "success": False,
                    "error": "Fund not found"
                }
            
            fund_id = fund['fund_id']
            
            # Get wallet information
            wallet_data = await self.crypto_manager.get_wallet(user_id, fund_id)
            
            if wallet_data:
                return {
                    "success": True,
                    "wallet": wallet_data
                }
            else:
                return {
                    "success": False,
                    "error": "Wallet not found"
                }
                
        except Exception as e:
            logger.error(f"Error getting wallet info for user {user_id}: {e}")
            return {
                "success": False,
                "error": f"Error getting wallet info: {str(e)}"
            }

File: frontend_dist/main-app/src/services/fund/fund-manager.ts
// frontend_dist/main-app/src/services/fund/fund-manager.ts
import { getLogger } from '../../boot/logging';
import { TokenManager } from '@trading-app/auth';
import { FundClient } from '@trading-app/api';
import { FundProfile, TeamMember, CreateFundProfileRequest, UpdateFundProfileRequest } from '../../types';
import { toastService } from '../notification/toast-service';

export class FundManager {
  private logger = getLogger('FundManager');
  private fundApi: FundClient;
  private tokenManager: TokenManager;

  constructor(fundApi: FundClient, tokenManager: TokenManager) {
    this.fundApi = fundApi;
    this.tokenManager = tokenManager;
    this.logger.info('FundManager initialized');
  }

  /**
   * Creates a new fund profile
   */
  async createFundProfile(fundData: CreateFundProfileRequest): Promise<{ 
    success: boolean; 
    fundId?: string;
    error?: string; 
  }> {
    if (!this.tokenManager.isAuthenticated()) {
      toastService.error('You must be logged in to create a fund profile');
      return { success: false, error: 'Not authenticated' };
    }

    try {
      const response = await this.fundApi.createFundProfile(fundData);
      
      if (response.success && response.fundId) {
        toastService.success(`Fund "${fundData.fundName}" created successfully`);
        return { 
          success: true, 
          fundId: response.fundId 
        };
      } else {
        toastService.error(response.error || 'Failed to create fund profile');
        return { 
          success: false, 
          error: response.error || 'Unknown error'
        };
      }
    } catch (error: any) {
      this.logger.error('Fund profile creation failed', error);
      toastService.error(`Failed to create fund profile: ${error.message}`);
      return { 
        success: false, 
        error: error.message || 'Unknown error'
      };
    }
  }

    
  /**
   * Retrieves the current user's fund profile
   */
  async getFundProfile(): Promise<{ 
    success: boolean; 
    fund?: FundProfile;
    error?: string; 
  }> {
    if (!this.tokenManager.isAuthenticated()) {
      return { success: false, error: 'Not authenticated' };
    }

    try {
      const response = await this.fundApi.getFundProfile();
      
      if (response.success && response.fund) {
        // Transform the API response to match our expected FundProfile structure
        // Use 'any' type for the API data since its structure differs from FundProfile
        const apiData: any = response.fund;
        
        // Create a properly formatted FundProfile object
        const formattedFund: FundProfile = {
          id: apiData.fund_id,
          userId: apiData.user_id,
          fundName: apiData.fundName,
          legalStructure: apiData.legalStructure,
          location: apiData.location,
          yearEstablished: apiData.yearEstablished,
          aumRange: apiData.aumRange,
          investmentStrategy: apiData.investmentStrategy,
          profilePurpose: apiData.profilePurpose || [],
          otherPurposeDetails: apiData.otherPurposeDetails,
          teamMembers: (apiData.team_members || []).map((member: any) => ({
            id: member.team_member_id,
            firstName: member.firstName,
            lastName: member.lastName,
            role: member.role,
            yearsExperience: member.yearsExperience,
            education: member.education,
            currentEmployment: member.currentEmployment,
            investmentExpertise: member.investmentExpertise,
            birthDate: member.birthDate,
            linkedin: member.linkedin,
          })),
          activeAt: apiData.active_at
        };
        
        return { 
          success: true, 
          fund: formattedFund 
        };
      } else {
        // Not finding a fund is a legitimate case, not an error
        return { 
          success: response.success,
          error: response.error
        };
      }
    } catch (error: any) {
      this.logger.error('Failed to fetch fund profile', error);
      return { 
        success: false, 
        error: error.message || 'Unknown error'
      };
    }
  }
  
  /**
   * Updates an existing fund profile
   */
  async updateFundProfile(updates: UpdateFundProfileRequest): Promise<{ 
    success: boolean;
    error?: string; 
  }> {
    if (!this.tokenManager.isAuthenticated()) {
      toastService.error('You must be logged in to update your fund profile');
      return { success: false, error: 'Not authenticated' };
    }

    try {
      const response = await this.fundApi.updateFundProfile(updates);
      
      if (response.success) {
        toastService.success('Fund profile updated successfully');
        return { success: true };
      } else {
        toastService.error(response.error || 'Failed to update fund profile');
        return { 
          success: false, 
          error: response.error || 'Unknown error'
        };
      }
    } catch (error: any) {
      this.logger.error('Fund profile update failed', error);
      toastService.error(`Failed to update fund profile: ${error.message}`);
      return { 
        success: false, 
        error: error.message || 'Unknown error'
      };
    }
  }
}